src/chain/shared_state.rs

```diff
@@ -292,3 +319,99 @@ impl Debug for SectionProofChain {
         )
     }
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::{chain::SectionInfo, BlsPublicKey, FullId, Prefix, XorName};
+    use std::collections::{BTreeSet, HashMap};
+    use std::str::FromStr;
+    use unwrap::unwrap;
+
+    fn gen_section_info(pfx: Prefix<XorName>) -> SectionInfo {
+        let sec_size = 5;
+        let mut members = BTreeSet::new();
+        for _ in 0..sec_size {
+            let id = FullId::within_range(&pfx.range_inclusive());
+            let _ = members.insert(*id.public_id());
+        }
+        unwrap!(SectionInfo::new(members, pfx, None))
+    }
+
+    fn gen_pk(pfx: Prefix<XorName>) -> BlsPublicKey {
+        BlsPublicKey::from_section_info(&gen_section_info(pfx))
+    }
+
+    // start_pfx: the prefix of our section as string
+    // updates: the prefixes of the sections we update the keys for, in sequence; every entry in
+    //          the vector will get its own key
+    // expected: vec of pairs (prefix, index)
+    //           the prefix is the prefix of the section whose key we check
+    //           the index is the index in the `updates` vector, which should have generated the
+    //           key we expect to get for the given prefix
+    fn update_keys_and_check(start_pfx: &str, updates: Vec<&str>, expected: Vec<(&str, usize)>) {
+        let keys_to_update = updates
+            .into_iter()
+            .map(|pfx_str| {
+                let pfx = unwrap!(Prefix::<XorName>::from_str(pfx_str));
+                (pfx, gen_pk(pfx))
+            })
+            .collect::<Vec<_>>();
+        let expected_keys = expected
+            .into_iter()
+            .map(|(pfx_str, index)| {
+                let pfx = unwrap!(Prefix::<XorName>::from_str(pfx_str));
+                (pfx, keys_to_update[index].1.clone())
+            })
+            .collect::<HashMap<_, _>>();
+
+        let start_section = gen_section_info(unwrap!(Prefix::from_str(start_pfx)));
+        let mut state = SharedState::new(start_section);
+
+        for (pfx, key) in keys_to_update {
+            state.update_their_keys(pfx, key);
+        }
+
+        assert_eq!(state.get_their_keys(), &expected_keys);
+    }
+
+    #[test]
+    fn single_prefix_multiple_updates() {
+        update_keys_and_check("0", vec!["1", "1", "1", "1"], vec![("1", 3)]);
```

Comment:
The fact that we keep the last that we see suggest that we have to make sure not to add a old one after the new one.
i think I mention this somewhere, but we will probably need to care about version number at some point.
(Nothing to do for this PR.)